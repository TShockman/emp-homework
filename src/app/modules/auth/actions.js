export const NAME = 'auth';

// lifecycle action names for a login request
export const LOGIN_REQUESTED = `${NAME}/LOGIN_REQUESTED`;
export const LOGIN_FULFILLED = `${NAME}/LOGIN_FULFILLED`;
export const LOGIN_REJECTED = `${NAME}/LOGIN_REJECTED`;
// NOTE: currently the rejected action is unused for simplicity sake
// since logging in always succeeds anyways. In a fuller implementation,
// LOGIN_REJECTED would be dispatched if the service request failed, the
// auth reducer would update the auth state to signify this, and the Login
// component would handle displaying relevant error text to the user.