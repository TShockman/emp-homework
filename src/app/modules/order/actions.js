export const NAME = 'order';

// action names for get orders flow
export const GET_ORDERS_REQUESTED = `${NAME}/GET_ORDERS_REQUESTED`;
export const GET_ORDERS_FULFILLED = `${NAME}/GET_ORDERS_FULFILLED`;
