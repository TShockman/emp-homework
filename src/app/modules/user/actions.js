export const NAME = 'user';

export const USER_REQUESTED = `${NAME}/USER_REQUESTED`;
export const USER_FULFILLED = `${NAME}/USER_FULFILLED`;
