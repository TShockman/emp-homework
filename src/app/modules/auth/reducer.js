import {Record} from 'immutable';
import {LOGIN_FULFILLED} from './actions';

// auth state is an immutable record holding only a user id
export const AuthState = Record({
  id: null
});

const initialState = new AuthState();

// on user login, update the state to hold the confirmed id
export default (state = initialState, action) => {
  switch (action.type) {
    case LOGIN_FULFILLED:
      return state.set('id', action.id);
    default: {
      return state;
    }
  }
}