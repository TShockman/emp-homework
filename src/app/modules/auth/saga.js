import {all, fork, select, call, put, takeLatest, getContext} from 'redux-saga/effects';
import {LOGIN_FULFILLED, LOGIN_REQUESTED} from './actions';
import EmpaticaService from '../../services/EmpaticaService';

const empaticaService = EmpaticaService.instance;

// attempt to login
function * loginUser({username, password}) {
  const user = {username, password};
  console.log('here')
  console.log(user)
  const result = yield call(empaticaService.login, user);

  if (result) {
    yield put({type: LOGIN_FULFILLED, id: result.id, username});
    yield call(alert, "Success!");
  } else {
    console.log("Error logging in.")
    yield call(alert, "Failure.");
    // TODO: handle error logging in
  }
}

export default function * rootSaga () {
  yield all([
    fork(takeLatest, LOGIN_REQUESTED, loginUser)
  ]);
}