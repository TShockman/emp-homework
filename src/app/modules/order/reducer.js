import {Map} from 'immutable';
import {GET_ORDERS_FULFILLED} from './actions';

// order state is a map from user id to list of order
const initialState = new Map();

export default (state = initialState, action) => {
  switch (action.type) {
    case GET_ORDERS_FULFILLED:
      return state.merge({[action.userId]: action.orders});
    default: {
      return state;
    }
  }
}