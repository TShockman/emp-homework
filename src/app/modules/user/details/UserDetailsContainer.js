import UserDetails from './UserDetails';
import {connect} from 'react-redux';
import {USER_REQUESTED} from '../actions';
import {selectUserState} from '../selectors';
import {selectAuthState} from '../../auth/selectors';
import {push} from 'redux-little-router';

function mapStateToProps(state) {
  const {user} = selectUserState(state);
  const {id} = selectAuthState(state);
  return {
    user,
    id
  };
}

function mapDispatchToProps(dispatch) {
  return {
    requestUser: ({id}) => dispatch({
      type: USER_REQUESTED,
      id
    }),
    push: href => dispatch(push(href))
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(UserDetails);
