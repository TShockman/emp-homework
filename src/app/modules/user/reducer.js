import {Record} from 'immutable';
import {USER_FULFILLED} from './actions';

export const UserState = Record({
  user: null
});

const initialState = new UserState();

export default (state = initialState, action) => {
  switch (action.type) {
    case USER_FULFILLED:
      return state.set('user', action.user);
    default: {
      return state;
    }
  }
}