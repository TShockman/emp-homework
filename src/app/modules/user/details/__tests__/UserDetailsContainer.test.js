/* eslint-disable import/first */
jest.mock('../UserDetails', () => 'UserDetails');

import React from 'react';
import renderer from 'react-test-renderer';
import {createStore} from 'redux';
import {Provider} from 'react-redux';
import {AuthState} from '../../../auth/reducer';
import {NAME as AUTH_NAME} from '../../../auth/actions';
import userReducer, {UserState} from '../../reducer';
import {NAME} from '../../actions';

import UserDetailsContainer from '../index';

describe('UserDetailsContainer', () => {
  it('Consistently provides UserDetails with correct props', () => {
    const store = createStore(userReducer, {
      [AUTH_NAME]: new AuthState({id: 123}),
      [NAME]: new UserState({
        user: 'mock user state user'
      })
    });
    const tree = renderer.create(
        <Provider store={store}>
          <UserDetailsContainer/>
        </Provider>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
