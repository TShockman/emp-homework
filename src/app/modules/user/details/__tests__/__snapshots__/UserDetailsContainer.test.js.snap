// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserDetailsContainer Consistently provides UserDetails with correct props 1`] = `
<UserDetails
  id={123}
  push={[Function]}
  requestUser={[Function]}
  user="mock user state user"
/>
`;
